contents/scalability/scalable-development.md

```diff
@@ -0,0 +1,63 @@
+
+Performance bottlenecks ([resource](https://www.apicasystem.com/blog/5-common-performance-bottlenecks/))
+
+Non-Relational Database ([resource](http://www.jamesserra.com/archive/2015/08/relational-databases-vs-non-relational-databases/)) ([resource](https://www.pluralsight.com/blog/software-development/relational-non-relational-databases))
+
+## Get More Machines
+
+Buy more machines to host your server. This is called horizontal scaling. This method can help improve scalability by a huge factor, as each additional unit of processing power and RAM costs the same, no matter how many machines you buy. As this is a promising way to increase scalability, It is no surprise why large services such as Google and Facebook use thousands of computer servers just to provide service to all their users.
+
+However, to fully reap the benefits of multiple machines, you will need to learn various techniques to modify your server infrastructure. In most cases you cannot simply run multiple instances of the server and expect it to work.
+
+Here are some tips to get started with horizontal scaling:
+- Learn how to utilize MapReduce on [Hadoop](https://www.tutorialspoint.com/hadoop/index.htm). The Hadoop ecosystem is so versatile and widely-used that there are companies offering Hadoop as a service, so you do not have to set up the server yourself
+- Host your server on [Google App Engine](https://cloud.google.com/appengine/docs). Despite being more expensive than Amazon Web Services and DigitalOcean, it offers very scalable infrastructure for your server.
+- Run a server with multiple nodes with MPI. [Open MPI](https://www.open-mpi.org/) is an open source message-passing library that can be used to send data between the nodes in your server. You might have to study [cloud computing infrastructure](http://whatiscloud.com/basic_concepts_and_terminology/it_resource) in order to make use of this.
+- Study various scalable architecture for servers ([resource](http://srinathsview.blogspot.sg/2011/10/list-of-known-scalable-architecture.html)).
+
+# But wait!
+
+Before you start tweaking your server to make it scalable, be sure to avoid the following pitfalls:
+- Do not optimize prematurely or unnecessarily: your "smart" tweak might save your server 3 seconds a day in return for making your life a living hell when you try to debug it 3 months later.
+- Do not start without a performance check: if processing an image seems to take a long time, you might spend days optimizing the processing function, only to find 99% of the time is spent sending it from one server to another.
+- Do not forget the goal: deliver the product first. There is no point hosting a server for millions of people when you cannot even hit a hundred users. Get your users first, then start worrying about scalability.
```

Comment:
Does this kind of contradict your main point (scalability is important, even in early stages)?